File: Blind 75/01 - Arrays and Hashing/JavaScript/128-longest-consecutive-sequence.js
const { performance } = require("perf_hooks");

class Solution {
  method(nums) {
    const seen = new Set(nums);
    let longestSequence = 0;
    for (const num of seen) {
      if (seen.has(num - 1)) continue;

      let count = 1;
      while (seen.has(num + count)) {
        count++;
      }
      longestSequence = Math.max(longestSequence, count);
    }

    return longestSequence;
  }

  reference(nums) {
    let len = nums.length;
    if (!len) {
      return 0;
    }
    const set = new Set(nums);
    let max = 0;

    for (let i = 0; i < nums.length; i++) {
      const num = nums[i];

      if (set.has(num - 1)) {
        continue;
      }

      let currentMax = 1;
      while (set.has(num + currentMax)) {
        currentMax++;
      }

      if (currentMax > max) {
        max = currentMax;
      }
      if (max > len / 2) {
        break;
      }
    }

    return max;
  }

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [100, 4, 200, 1, 3, 2],
  [0, 3, 7, 2, 5, 8, 4, 6, 0, 1],
];
test.quantify(testCases);

File: Blind 75/01 - Arrays and Hashing/128-longest-consecutive-sequence.py
'''
Given an unsorted array of integers nums, return the length of the longest consecutive elements sequence.

You must write an algorithm that runs in O(n) time.
'''

from time import time
from typing import List


class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        longest = 0
        seen = set(nums)
        for n in nums:
            if n - 1 not in seen:
                # Loop only at the beginning of a sequence
                curr = n
                count = 0
                while curr in seen:
                    count += 1
                    curr += 1
                longest = max(longest, count)

        return longest

    def reference(self, nums: List[int]) -> List[int]:
        numSet = set(nums)
        longest = 0

        for n in nums:
            # Check if its the start of a sequence
            if (n - 1) not in numSet:
                length = 1
                while (n + length) in numSet:
                    length += 1
                longest = max(length, longest)
        return longest

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.longestConsecutive(case))
                else:
                    self.longestConsecutive(case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [[100, 4, 200, 1, 3, 2], [0, 3, 7, 2, 5, 8, 4, 6, 0, 1]]
    test.quantify(test_cases)
